libraries/apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/introspection/IntrospectionSchema.kt

```diff
@@ -238,7 +238,16 @@ fun BufferedSource.toIntrospectionSchema(origin: String = ""): IntrospectionSche
 
   return try {
     val introspectionSchemaEnvelope = json.decodeFromString(IntrospectionSchemaEnvelope.serializer(), this.readUtf8())
-    introspectionSchemaEnvelope.data ?: introspectionSchemaEnvelope.__schema?.let { IntrospectionSchema(it) }
+    introspectionSchemaEnvelope.data ?: introspectionSchemaEnvelope.__schema?.let { schema ->
+      IntrospectionSchema(IntrospectionSchema.Schema(
+          queryType = schema.queryType,
+          mutationType = schema.mutationType,
+          subscriptionType = schema.subscriptionType,
+          types = schema.types,
+          // Old introspection json (pre `April2016`) may not have the `locations` field, in which case the list will be empty, which is invalid. Exclude those directives.
+          directives = schema.directives.filter { directive -> directive.locations.isNotEmpty() }
```

Comment:
Have you ever seen such a schema? 

This is certainly the way to go for 3.x because [we ignore unknown directives](https://github.com/apollographql/apollo-kotlin/blob/05f2a3295517fb7a64c2259c38b453a37f3c4d19/libraries/apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/internal/ValidationCommon.kt#L133). But 4.x should certainly require a post 2016 schema

Reply:
> Have you ever seen such a schema?
No, except in the json of our own tests (this didn't cause any issue because the whole `directives` list was ignored)

Reply:
I can remove the `= emptyList()` part on directives for v4 then - this will throw at parse time.

Reply:
Yea I think that's OK. It's not the best error but it should not happen too much hopefully 🤞